### PR TITLE
feat(express-tooling): inject custom getTokenInfo

### LINKED ISSUE #219: Custom getTokenInfo for authenticationMiddleware
Hi! my team started using `authenticationMiddleware` at work. We noticed that `getTokenInfo` was not using HTTP keep-alive and leaving thousands of `TIME_WAIT` TCP connections.

Because it is a crucial RPC call, I'd like to have the following reliability measures for `getTokenInfo` 

- HTTP keep-alive (using HTTP agent)
- timeout
- retry
- metrics for monitoring
- circuit breaker (not sure if it makes sense for now)

But if `authenticationMiddleware` supports all of them directly, it will have too many options that are not directly related to authentication or the middleware. So I'd like to propose to make `getTokenInfo` injectable.

```ts
type GetTokenInfo<T> = (tokenInfoUrl: string, accessToken: string, logger?: Logger) => Promise<Token<T>>;

type AuthenticationMiddlewareOptions = {
  tokenInfoEndpoint: string,
  logger?: Logger,
  onNotAuthenticatedHandler?: onNotAuthenticatedHandler,
  publicEndpoints?: string[],
  getTokenInfo?: GetTokenInfo<{}>
};
```

If the `getTokenInfo` is not provided in the option, the middleware can use the default implementation.

I appreciate your feedback.

### PR DESCRIPTION
**Pull request by shuhei**
close #219

### CODE REVIEW

File: src/types/AuthenticationMiddlewareOptions.ts [hunk 1]
```diff
@@ -1,13 +1,14 @@
 import { NextFunction, Request, Response } from 'express';
-import { Logger } from '.';
+import { Logger, GetTokenInfo } from '.';
 
 type onNotAuthenticatedHandler = (request: Request, resonse: Response, next: NextFunction, logger: Logger) => void;
 
-type AuthenticationMiddlewareOptions = {
+type AuthenticationMiddlewareOptions<T = {}> = {
```
Comment (shuhei) [inline src/types/AuthenticationMiddlewareOptions.ts]: Is this type parameter fine? I'm not an expert of TypeScript...

File: src/types/GetTokenInfo.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import { Logger, Token } from '.';
+
+type GetTokenInfo<T> = (tokenInfoUrl: string, accessToken: string, logger?: Logger) => Promise<Token<T>>;
```
Comment (bzums) [inline src/types/GetTokenInfo.ts]: We should also use this type for our implementation of `getTokenInfo`: https://github.com/zalando-incubator/authmosphere/blob/master/src/oauth-tooling.ts#L156

Then that would look like:

```ts
const getTokenInfo: GetTokenInfo = (tokenInfoUrl: string, accessToken: string, logger?: Logger) => {
  // ...
```

To make this possible, the type has to provide a default generic type:

```ts
type GetTokenInfo<T = {}> = (tokenInfoUrl: string, accessToken: string, logger?: Logger) => Promise<Token<T>>;
```

File: src/types/AuthenticationMiddlewareOptions.ts (hunk 1 above)
Comment (bzums) [reply src/types/AuthenticationMiddlewareOptions.ts]: I think you can remove the generic type annotation here if you add it to `GetTokenInfo` (see next comment).

**Comment by bzums** [comment]
@shuhei do you need any further support here? Happy to help!

File: src/types/GetTokenInfo.ts (hunk 2 above)
Comment (shuhei) [reply src/types/GetTokenInfo.ts]: Ah, got it! I tried to, but couldn’t  make it because I was missing `<T = {}>`.

**Comment by bzums** [comment]
@shuhei To the PR in general: since we now also would like to monitor tokeninfo endpoint response times we would like to see this going into the master. I did the adaptions I suggested in https://github.com/zalando-incubator/authmosphere/tree/219-custom-get-token-info (rest of the code is copy-pasted from your branch) so we could close this PR and open a new one if that is fine for you?

Regarding the return type of `GetTokenInfo`: you are definitely right, it would be nice to have the status code in the returned object. I would propose to open a ticket and discuss details there. WDYT?

**Comment by shuhei** [comment]
@bzums Sure, it's totally fine to close this PR and have a ticket for `GetTokenInfo` type.

**Comment by bzums** [comment]
@shuhei great, added you as reviewer to the PR in case you are interested in the changeset (though there is not much difference to yours ;))

**Pull request closed without merging**